Reuse hash in instance.Push to reduce memory allocations

mdisibio opened the pull request:
**What this PR does**:
Small optimization to ingester instance.getOrCreateTrace method to reuse hash instead of allocating a new one inside TokenForTraceID. This eliminates 1 allocation per Push.   Basic benchmarks were added:

Before:
```
$ go test -bench . -benchmem
goos: darwin
goarch: amd64
pkg: github.com/grafana/tempo/modules/ingester
BenchmarkInstancePush-12                 	 1817228	       650 ns/op	     197 B/op	       4 allocs/op
BenchmarkInstancePushExistingTrace-12    	 7747898	       152 ns/op	      47 B/op	       1 allocs/op
BenchmarkInstanceFindTraceByID-12        	   90445	     12661 ns/op	   11035 B/op	     209 allocs/op
PASS
ok  	github.com/grafana/tempo/modules/ingester	3.801s
```

After:
```
$ go test -bench . -benchmem       
goos: darwin
goarch: amd64
pkg: github.com/grafana/tempo/modules/ingester
BenchmarkInstancePush-12                 	 1887462	       644 ns/op	     190 B/op	       3 allocs/op
BenchmarkInstancePushExistingTrace-12    	 8306944	       138 ns/op	      40 B/op	       0 allocs/op
BenchmarkInstanceFindTraceByID-12        	   82716	     12498 ns/op	   11035 B/op	     208 allocs/op
PASS
ok  	github.com/grafana/tempo/modules/ingester	3.786s
```

**Which issue(s) this PR fixes**:
n/a

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

joe-elliott commented on modules/ingester/instance.go:
```diff
@@ -61,6 +63,8 @@ type instance struct {
 	limiter            *Limiter
 	wal                *tempodb_wal.WAL
+
+	hash hash.Hash32
```
it occurs to me that this might not be concurrency safe.  if it's not we can't make this change.  apologies for sending you down the wrong path.

obviously the `CombineTraces` change is still fine b/c it is all executed in the same goroutine.

mdisibio replied:
Yes that was considered, but it seems like it is concurrency safe?   The only call chain is under the traces mutex lock via tokenForTraceID <- getOrCreateTrace <- Push.

joe-elliott replied:
Yup, you are correct.  It is protected by a mutex.

joe-elliott commented:
One other concern.  This change splits the hash generation method used in `FindTraceByID()` from the one used in `getOrCreateTrace()`.  I think we should change them to both use the same one.

mdisibio commented:
> One other concern. This change splits the hash generation method used in `FindTraceByID()` from the one used in `getOrCreateTrace()`. I think we should change them to both use the same one.

Updated and added the benchmark results above, showing the elimination of 1 alloc/op.
